crypto/src/mls/client/identities.rs

```diff
@@ -5,106 +5,120 @@ use crate::{
 use openmls::prelude::{Credential, SignaturePublicKey};
 use openmls_traits::types::SignatureScheme;
 use std::collections::HashMap;
+use std::sync::Arc;
+use async_lock::Mutex;
 
 /// In memory Map of a Client's identities: one per SignatureScheme.
 /// We need `indexmap::IndexSet` because each `CredentialBundle` has to be unique and insertion
 /// order matters in order to keep values sorted by time `created_at` so that we can identify most recent ones.
 #[derive(Debug, Clone)]
-pub(crate) struct ClientIdentities(HashMap<SignatureScheme, indexmap::IndexSet<CredentialBundle>>);
+pub(crate) struct ClientIdentities(Arc<Mutex<HashMap<SignatureScheme, indexmap::IndexSet<CredentialBundle>>>>);
 
 impl ClientIdentities {
     pub(crate) fn new(capacity: usize) -> Self {
-        Self(HashMap::with_capacity(capacity))
+        Self(Arc::new(Mutex::new(HashMap::with_capacity(capacity))))
     }
 
-    pub(crate) fn find_credential_bundle_by_public_key(
+    pub(crate) async fn find_credential_bundle_by_public_key(
         &self,
         sc: SignatureScheme,
         ct: MlsCredentialType,
         pk: &SignaturePublicKey,
-    ) -> Option<&CredentialBundle> {
-        self.0.get(&sc)?.iter().find(|c| {
+    ) -> Option<CredentialBundle> {
+        let credential_bundles_by_signature_schemes = self.0.lock().await;
+        let credential_bundles = credential_bundles_by_signature_schemes.get(&sc)?;
+        credential_bundles.iter().find(|c| {
             let ct_match = ct == c.credential.credential_type().into();
             let pk_match = c.signature_key.public() == pk.as_slice();
             ct_match && pk_match
-        })
+        }).cloned()
```

Comment:
Why do we clone bundles now?

Reply:
Couldn't figure out a way to return references out of the mutex; Is there one?